include/up-cpp/utils/CallbackConnection.h

```diff
@@ -118,79 +150,151 @@ struct [[nodiscard]] Connection {
+			}
+
+			notify_if_expired(std::move(callback));
+		}
+	}
 
 	///////////////////////////////////////////////////////////////////////////
-private:
 	/// @brief Workaround for std::shared_ptr requiring a public constructor
-	struct PrivateConstructToken {};
+	///
+	/// This will always compile out while also enlisting the compiler to
+	/// restrict access to constructors that take ome of these as a parameter.
+	class PrivateConstructToken {
+		// Allow establish() to construct our classes
+		friend Connection;
+		// Allow the CalleeHandle to construct a CallerHandle when a cleanup
+		// callback is provided.
+		friend Handle;
+		/// @brief Private constructor
+		PrivateConstructToken() = default;
+	};
 
-public:  // But actually still private thanks to the PCT
 	/// @brief Semi-private constructor. Use the static establish() instead.
-	Connection(std::shared_ptr<Callback> cb, PrivateConstructToken);
+	Connection(std::shared_ptr<Callback> cb, PrivateConstructToken)
+	    : callback_(cb) {}
 
 	// Connection is only ever available wrapped in a std::shared_ptr.
 	// It cannot be moved or copied directly without breaking the
 	// corresponding handle, so those constructors are disabled.
-	// clang-format off
 	Connection(Connection&&) = delete;
-	// clang-format on
 	Connection(const Connection&) = delete;
 	Connection& operator=(const Connection&) = delete;
 
 private:
 	// The CalleeHandle is allowed to use sever() to terminate the connection
 	friend Handle;
-	void sever();
+	void sever() {
+		sever_requested_ = true;
+		std::unique_lock lk(callback_mtx_);
+		sever_cv_.wait(lk, [this]() { return callback_.expired(); });
+	}
 
 	// Unblocks sever() if all callbacks have finished
 	// NOTE: Must be called from inside a block that has checked
 	// for !sever_requested_
-	// clang-format off
-	void notify_if_expired(std::weak_ptr<Callback>&& cb);
-	// clang-format on
+	void notify_if_expired(std::weak_ptr<Callback>&& cb) {
+		if (cb.expired()) {
+			sever_cv_.notify_one();
+		}
+	}
 
 	std::atomic<bool> sever_requested_{false};
-	// Note: would use std::atomic<std::weak_ptr<Callback>> if we had C++20
-	std::mutex callback_mtx_;
+	// Note: would use std::atomic<std::weak_ptr<Callback>> if we had C++20.
+	// It would replace the mutex, cv, and plain weak_ptr all in one object.
+	mutable std::mutex callback_mtx_;
+	std::condition_variable sever_cv_;
 	std::weak_ptr<Callback> callback_;
 };
 
+/// @brief Thrown if construction of one of the handles fails due to an invalid
+///        connection pointer.
+///
+/// This would only occur if std::make_shared<Connection> failed for some
+/// reason.
+struct BadConnection : public std::runtime_error {
+	template <typename... Args>
+	BadConnection(Args&&... args)
+	    : std::runtime_error(std::forward<Args>(args)...) {}
+};
+
 template <typename RT, typename... Args>
 struct [[nodiscard]] CalleeHandle {
 	using Conn = Connection<RT, Args...>;
 	using Callback = typename Conn::Callback;
```

Comment:
Removed. It does mean that the `typename` keyword becomes necessary, so instead of `shared_ptr<Callback>` it will read `shared_ptr<typename Conn::Callback>`.